AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -796,17 +798,142 @@ - (BOOL)shouldCalculateLayoutWithConstrainedSize:(ASSizeRange)constrainedSize pa
+  
+  // Invalidate the current calculated layout so in the next layout pass it get's recalculated
+  [self invalidateCalculatedLayout];
+  
+  // Walk the node's supernodes until a ancestor node is found that does not require layout
+  if ([_supernode __needsLayout]) {
+    ASDisplayNode *supernode = _supernode;
+    __instanceLock__.unlock();
+    [_supernode __layoutIfNeeded];
+    return;
+  }
+
+  // Layout all subviews starting from the first node that needs layout
+  [self __layout];
+  
+  __instanceLock__.unlock();
+}
+
+/**
+ * Called from [CALayer layoutSublayers:]: All subnode layout related code should happen in there
+ */
+- (void)__layout
+{
+  ASDisplayNodeAssertMainThread();
+  ASDN::MutexLocker l(__instanceLock__);
+  CGRect bounds = _threadSafeBounds;
+  
+  // Prevent zero bounds subnode adjustments
+  if (CGRectEqualToRect(bounds, CGRectZero)) {
+    // Performing layout on a zero-bounds view often results in frame calculations with negative sizes after applying
+    // margins, which will cause layoutThatFits: on subnodes to assert.
+    LOG(@"Warning: No size given for node before node was trying to layout itself: %@. Please provide a frame for the node.", self);
+    return;
+  }
+
+  // Node needs to be measured in case the bounds changed from the current calculated layout
+  [self calculateLayoutIfNecessaryWithBounds:bounds];
+  
+  // Handle placeholder layer creation in case the size of the node changed after the initial placeholder layer
+  // was created
+  if ([self _shouldHavePlaceholderLayer]) {
+    [self _setupPlaceholderLayerIfNeeded];
+  }
+  _placeholderLayer.frame = bounds;
+  
+  // Layout all subnodes based on the layout determined before
+  [self layout];
+  [self layoutDidFinish];
+}
+
+/*
+ * Executes a measure pass if necessary based on the current bounds as constrained size
+ */
+- (void)calculateLayoutIfNecessaryWithBounds:(CGRect)bounds
+{
+  // Setup new constrained and parent size based on bounds
+  ASSizeRange constrainedSize = ASSizeRangeMake(bounds.size);
+  CGSize parentSize = constrainedSize.max;
+  
+  // Check if we have to create a layout before the layout pass can happen
+  CGSize calculatedDisplayNodeLayoutSize = CGSizeZero;
+  BOOL needsCalculateLayout = NO;
+  {
+    ASDN::MutexLocker l(__instanceLock__);
+    calculatedDisplayNodeLayoutSize = _calculatedDisplayNodeLayout->layout.size;
+    needsCalculateLayout = [self shouldCalculateLayoutWithConstrainedSize:constrainedSize parentSize:parentSize];
+  }
+  
+  // Check if we can reuse the calculated layout for the layout pass
+  if (CGSizeEqualToSize(calculatedDisplayNodeLayoutSize, CGSizeZero) == NO &&
+      CGSizeEqualToSize(calculatedDisplayNodeLayoutSize, bounds.size) == YES &&
+      needsCalculateLayout == NO
+      ) {
+    // We can use the calculated layout for the layout subnodes so just return
+    return;
+  }
+  
+  // Cancel layout transition that is in flight
+  [self cancelLayoutTransition];
```

Comment:
I agree. I changed to hold the `__instanceLock__`  to the whole method as we also create the layout transition within this method and assign it to an instance method. That said all methods we call within within will grab the lock so there would be some improvements possible.